src/mqtt_client.rs

```diff
@@ -288,15 +243,27 @@ where
         Ok(())
     }
 
-    fn handle_packet<F>(&mut self, f: &mut F) -> Result<(), Error<N::Error>>
-    where
-        for<'a> F: FnMut(&Self, &'a str, &[u8], &[Property<'a>]),
-    {
-        let packet =
-            ReceivedPacket::parse_message(&self.packet_reader).map_err(|e| Error::Protocol(e))?;
+    fn connect_socket(&mut self) -> Result<(), Error<N::Error>> {
+        let socket = self.socket.as_mut().unwrap();
```

Comment:
Instead of unwrapping here, this could be the place that gets the socket handle.